nixos/modules/security/pwquality/pwquality-conf-format.nix

```
@@ -0,0 +1,274 @@
+          length in some form.
+        '';
+      };
+
+      enforcing = lib.mkOption {
+        type = types.nullOr types.ints.unsigned;
+        default = null;
+        example = 1;
+        description = ''
+          If nonzero, reject the password if it fails the checks,
+          otherwise only print the warning. This setting applies only to
+          the `pam_pwquality` module and possibly other applications that
+          explicitly change their behavior based on it. It does not affect
+          {manpage}`pwmake(1)` and {manpage}`pwscore(1)`
+        '';
+      };
+
+      badwords = lib.mkOption {
+        type = types.nullOr (types.listOf types.str);
+        default = null;
+        example = [
+          "password"
+          "pass"
+          "hunter42"
+        ];
+        description = ''
+          A list of words that must not be contained in the password.
+          These are additional words to the cracklib dictionary check.
+          This setting can be also used by applications to emulate the
+          gecos check for user accounts that are not created yet.
+        '';
+      };
+
+      dictpath = lib.mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = "/usr/local/lib/pw_dict";
+        description = ''
+          Path to the cracklib dictionaries. Default is to use the
+          cracklib default.
+        '';
+      };
+
+      retry = lib.mkOption {
+        type = types.nullOr types.ints.unsigned;
+        default = null;
+        example = 1;
+        description = ''
+          Prompt user at most *N* times before returning with error.
+        '';
+      };
+
+      enforce_for_root = lib.mkOption {
+        type = types.nullOr types.bool;
+        default = null;
+        example = false;
+        description = ''
+          The module will return error on failed check even if the user
+          changing the password is root. This option is off by default
+          which means that just the message about the failed check is
+          printed but root can change the password anyway. Note that root
+          is not asked for an old password so the checks that compare the
+          old and new password are not performed.
+        '';
+      };
+
+      local_users_only = lib.mkOption {
+        type = types.nullOr types.bool;
+        default = null;
+        example = true;
+        description = ''
+          The module will not test the password quality for users that are
+          not present in the {file}`/etc/passwd` file. The module still
+          asks for the password so the following modules in the stack can
+          use the `use_authtok` ({manpage}`pam_unix(8)`) option.
+        '';
+      };
+    };
+  };
+
+  generate =
+    settings:
+    (lib.concatMapStrings (x: x + "\n") (
+      lib.mapAttrsToList (
+        name: value:
+        if (builtins.typeOf value) == "bool" then name else "${name} = ${builtins.toString value}"
```

- `toString` is in the root scope
- its list behavior is a bit implicit
- bool processing is a bit spread out otherwise

```suggestion
        # toString will concatenate lists of strings with space: " "
        if value == true then name else assert value != false; "${name} = ${toString value}"
```

---

Alternatively or in addition, a unit test would make this easier to work with.